File: basic_python/practices/07. Tree/Invert_BinaryTree/invert.py
from typing import Optional
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def build_tree(self, arr):
        if not arr: return None
        root = TreeNode(arr[0])
        queue = deque([root])
        i = 1
        while queue and i < len(arr):
            node = queue.popleft()
            if i < len(arr): 
                node.left = TreeNode(arr[i]) if arr[i] is not None else None
                if node.left: queue.append(node.left)
                i += 1
            if i < len(arr): 
                node.right = TreeNode(arr[i]) if arr[i] is not None else None
                if node.right: queue.append(node.right)
                i += 1
        return root

    def level_order(self, root):
        if not root: return []
        queue = deque([root])
        result = []
        while queue:
            node = queue.popleft()
            result.append(node.val)
            if node.left: queue.append(node.left)
            if node.right: queue.append(node.right)
        return result

    def print_level_visual(self, root):
        """Print tree in a visually balanced form."""
        if not root:
            return

        from collections import deque
        
        # Step 1: BFS to get all levels
        levels = []
        queue = deque([(root, 0)])
        current_level = 0
        current_nodes = []

        while queue:
            node, level = queue.popleft()
            if level != current_level:
                levels.append(current_nodes)
                current_nodes = []
                current_level = level
            current_nodes.append(node.val if node else None)

            if node:
                queue.append((node.left, level+1))
                queue.append((node.right, level+1))
        levels.append(current_nodes)

        # Step 2: print with spacing
        max_level = len(levels)
        max_width = 2 ** max_level

        for depth, level_nodes in enumerate(levels):
            spacing = max_width // (2 ** depth)
            line = (" " * (spacing // 2)).join(
                [' ' if v is None else str(v) for v in level_nodes]
            )
            print(" " * (spacing // 2) + line)

    def print_top_down(self, root):
        """Print the tree in a balanced top-down (triangle) format."""
        if not root:
            print("<empty tree>")
            return
        
        # Get tree height
        def height(node):
            if not node:
                return 0
            return 1 + max(height(node.left), height(node.right))
        
        h = height(root)
        max_width = 2 ** h  # widest number of slots

        # BFS levels (including None children to preserve structure)
        from collections import deque
        result = []
        queue = deque([root])
        for _ in range(h):
            level = []
            next_q = deque()
            while queue:
                node = queue.popleft()
                level.append("" if node is None else str(node.val))
                if node:
                    next_q.append(node.left)
                    next_q.append(node.right)
                else:
                    next_q.append(None)
                    next_q.append(None)
            result.append(level)
            queue = next_q

        # Print top-down with spacing
        for depth, level in enumerate(result):
            spacing = max_width // (2 ** depth)
            line = (" " * (spacing // 2)).join(
                val.center(2) for val in level
            )
            print(" " * (spacing // 2) + line)

# Main
class invertTree:
    def use_BFS(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: return None
        queue = deque([root])
        while queue:
            node = queue.popleft()
            node.left, node.right = node.right, node.left
            if node.left: queue.append(node.left)
            if node.right: queue.append(node.right)
        return root

# Usage
if __name__ == "__main__":

    def parse_value(x):
        if x.lower() == "none":
            return None
        return int(x)

    raw = input("Input your array here: ")

    # clean input
    raw = raw.replace("[", "").replace("]", "").replace(",", " ")
    arr = [parse_value(x) for x in raw.split()]

    tree = TreeNode()
    root = tree.build_tree(arr)

    print("\nInput Tree:")
    tree.print_top_down(root)
    print("Input level order:", tree.level_order(root))

    sol = invertTree()
    inverted = sol.use_BFS(root)

    print("\nInverted Tree:")
    tree.print_top_down(inverted)
    print("Output level order:", tree.level_order(inverted))

File: basic_python/practices/07. Tree/Invert_BinaryTree/invert.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// ---------- build tree (level-order, allow None) ----------
TreeNode* build_tree(const vector<optional<int>>& arr) {
    if (arr.empty() || !arr[0].has_value())
        return nullptr;

    TreeNode* root = new TreeNode(arr[0].value());
    queue<TreeNode*> q;
    q.push(root);

    int i = 1;
    while (!q.empty() && i < arr.size()) {
        TreeNode* node = q.front();
        q.pop();

        // left child
        if (i < arr.size() && arr[i].has_value()) {
            node->left = new TreeNode(arr[i].value());
            q.push(node->left);
        }
        i++;

        // right child
        if (i < arr.size() && arr[i].has_value()) {
            node->right = new TreeNode(arr[i].value());
            q.push(node->right);
        }
        i++;
    }
    return root;
}

// ---------- level order ----------
vector<int> level_order(TreeNode* root) {
    vector<int> result;
    if (!root) return result;

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* node = q.front(); q.pop();
        result.push_back(node->val);
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    return result;
}

// ---------- height ----------
int height(TreeNode* node) {
    if (!node) return 0;
    return 1 + max(height(node->left), height(node->right));
}

// ---------- print top-down ----------
void print_top_down(TreeNode* root) {
    if (!root) {
        cout << "<empty tree>\n";
        return;
    }

    int h = height(root);
    int max_width = 1 << h; // 2^h

    // BFS including nullptr placeholders
    vector<vector<string>> levels;
    queue<TreeNode*> q;
    q.push(root);

    for (int lvl = 0; lvl < h; lvl++) {
        vector<string> level;
        queue<TreeNode*> next_q;

        while (!q.empty()) {
            TreeNode* node = q.front(); q.pop();

            if (node)
                level.push_back(to_string(node->val));
            else
                level.push_back("");

            if (node) {
                next_q.push(node->left);
                next_q.push(node->right);
            } else {
                next_q.push(nullptr);
                next_q.push(nullptr);
            }
        }

        levels.push_back(level);
        q = next_q;
    }

    // pretty-print
    for (int depth = 0; depth < levels.size(); depth++) {
        int spacing = max_width >> depth;

        cout << string(spacing / 2, ' ');
        for (string v : levels[depth]) {
            string cell = v;
            if (cell == "") cell = " ";
            cout << cell << string(spacing / 2, ' ');
        }
        cout << "\n";
    }
}

// ---------- invert tree BFS ----------
TreeNode* invertTree(TreeNode* root) {
    if (!root) return nullptr;
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* node = q.front(); q.pop();
        swap(node->left, node->right);
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    return root;
}

// ---------- main ----------
int main() {
    cout << "Input your array here: ";
    string raw;
    getline(cin, raw);

    // clean input
    for (char& c : raw)
        if (c == '[' || c == ']' || c == ',') c = ' ';

    stringstream ss(raw);
    string token;

    vector<optional<int>> arr;

    while (ss >> token) {
        if (token == "None" || token == "none" || token == "NULL")
            arr.push_back(nullopt);
        else
            arr.push_back(stoi(token));
    }

    TreeNode* root = build_tree(arr);

    cout << "\nInput Tree:\n";
    print_top_down(root);

    vector<int> before = level_order(root);
    cout << "Input level order: ";
    for (int v : before) cout << v << " ";
    cout << "\n\n";

    root = invertTree(root);

    cout << "Inverted Tree:\n";
    print_top_down(root);

    vector<int> after = level_order(root);
    cout << "Output level order: ";
    for (int v : after) cout << v << " ";
    cout << "\n";

    return 0;
}

/* ---------- For C++ 17
Input your array here: 1 2 3 4 5 6 7

Input Tree:
    1    
  2  3  
 4 5 6 7 
Input level order: 1 2 3 4 5 6 7 

Inverted Tree:
    1    
  3  2  
 7 6 5 4 
Output level order: 1 3 2 7 6 5 4 

-------- */